### PR TITLE
[5.5] Add ability to exclude specific columns from the result collection of Eloquent query

### PR DESCRIPTION
**Pull request by AhmedEssamElNaggar**
I made a pull request for this before but it's been closed before I  can fix the code for the expected behavior that @sisve warned me about in the discussion. I couldn't change the previous pull request since It's been closed so I hope this is fine to make a pull request for this again.

Thanks.

### CODE REVIEW

**Comment by gabriel-caruso** [comment]
Rebase your fork, to avoid many commits 💯

**Comment by Dylan-DPC** [comment]
You haven't added tests yet for it

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -245,6 +245,22 @@ public function orWhere($column, $operator = null, $value = null)
...
+    public function exclude($columns)
+    {
+        $wantedColumns = array_diff(
+                $this->query->columns,
+                (array) $columns
+            );
+
+        return $this->select($wantedColumns);
+    }
+
```
Comment (tillkruss) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please add tests.

**Review by tillkruss — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 2]
```diff
@@ -245,6 +245,22 @@ public function orWhere($column, $operator = null, $value = null)
     }
 
     /**
+     * Exclude one or many columns from being in the result colelction.
```
Comment (dwightwatson) [inline src/Illuminate/Database/Eloquent/Builder.php]: Typo in `collection`.

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 1 above)
Comment (lilessam) [reply src/Illuminate/Database/Eloquent/Builder.php]: Ok I'm working on it now.

Comment (lilessam) [reply src/Illuminate/Database/Eloquent/Builder.php]: @tillkruss Can you guide me? I usually use PHPUnit for my laravel applications but I don't know how to do it to my forked repo.

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 2 above)
Comment (lilessam) [reply src/Illuminate/Database/Eloquent/Builder.php]: @dwightwatson Will fix it, Thanks.

**Comment by sisve** [comment]
It looks like this would fail for `User::exclude('email')->get();` since the get's signature include the parameters; `function get($columns = ['*'])`.

**Comment by sisve** [comment]
For reference; the previous pull request is at https://github.com/laravel/framework/pull/21681

**Comment by lilessam** [comment]
@sisve  I don't know if `User::select('email')->get();` will fail too?

What do you suggest for this case?

**Comment by gabriel-caruso** [comment]
@lilessam Do you want help to write some tests? I submit a PR in your fork and you merge it!

**Comment by lilessam** [comment]
@gabriel-caruso Yes my friend if you can help me to write tests it will be great. After that I'll rebase the repo to avoid many commits like you said before.

I'm waiting your help ! :)

**Comment by gabriel-caruso** [comment]
@lilessam If someone could help us as well, I'd be grateful. Most of `EloquentBuilderTest` are Mocks, and I'm not familiar with that :cry:

**Comment by sisve** [comment]
@lilessam It is my belief that this functionality cannot be done without first querying the database for all available columns. That is the only way you can go from a "blind" select-everything (`SELECT * FROM user`) query to selecting all the fields (`SELECT id, email, name FROM user`). The resulting list of columns can then be the basis of manipulation with your new exclusion functionality.

I think your previous PR was on the right track, but there were some cases that it didn't handle, like chaining several exclude calls.

**Comment by taylorotwell** [comment]
No plans on adding.

**Pull request closed without merging**